include/swift/Basic/Assertions.h

```diff
@@ -0,0 +1,188 @@
+//===--- Assertions.h - Assertion macros                               ----===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file provides three alternatives to the C/C++ standard `assert()` macro
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_BASIC_ASSERTIONS_H
+#define SWIFT_BASIC_ASSERTIONS_H
+
+#include <stdint.h>
+
+// Only for use in this header
+#if __has_builtin(__builtin_expect)
+#define ASSERT_UNLIKELY(expression) (__builtin_expect(!!(expression), 0))
+#else
+#define ASSERT_UNLIKELY(expression) ((expression))
+#endif
+
+// ================================ Mandatory Asserts ================================
+
+// `ASSERT(expr)`:
+// * is always compiled into the executable and
+// * always checks the condition, regardless of build or runtime settings.
+// This should be used for most assertions, which is why it
+// deserves the short name.  In particular, for simple checks
+// (e.g., validating that something is non-null), this is just as
+// fast as a disabled `CONDITIONAL_ASSERT`, so there's no point in
+// using the conditional form.
+//
+// You can globally replace `assert` with `ASSERT` in a piece of code
+// to have all your assertions enabled in all builds.  If you do this,
+// please do a little profiling first, just in case you have some checks
+// that are more expensive than you think.  You can switch those to
+// `CONDITIONAL_ASSERT` or `DEBUG_ASSERT` as needed.
+
+#define ASSERT(expr) \
+  do { \
+    if (ASSERT_UNLIKELY(!expr)) {			   \
+      ASSERT_failure(#expr, __FILE__, __LINE__, __func__); \
```

Comment:
You should use `__FILE_NAME__`, not `__FILE__`, here and in the other macros. That'll let you delete the path-removal logic in `ASSERT_failure()`, and much like `#fileID` on the Swift side, it'll reduce code size and avoid leaking details about the build environment into the binary.

Reply:
Addressed in #74581